src/backup_domain.rs

```diff
@@ -22,3 +22,143 @@ use crate::pac::BKP;
 pub struct BackupDomain {
     pub(crate) _regs: BKP,
 }
+
+
+macro_rules! write_drx {
+    ($self:ident, $drx:ident, $new:expr) => ({
+        let old: u16 = $self._regs.$drx.read().bits() as u16;
+        if old != $new {
+            unsafe {
+                $self._regs.$drx.write(|w| w.bits($new.into()));
+            }
+        }
+    })
+}
+
+macro_rules! read_drx {
+    ($self:ident, $drx:ident) => (
+        $self._regs.$drx.read().bits() as u16;
+    )
+}
+
+impl BackupDomain {
+    /// Read a 16-bit value from the Backup Data Register. The first argument
+    /// denotes the DRx register, i.e. 1 is DR1, up to DR42.
+    pub fn read_bkp_dr(&self, register: u8) -> u16 {
+        match register {
+            1  => read_drx!(self, dr1),
+            2  => read_drx!(self, dr2),
+            3  => read_drx!(self, dr3),
+            4  => read_drx!(self, dr4),
+            5  => read_drx!(self, dr5),
+            6  => read_drx!(self, dr6),
+            7  => read_drx!(self, dr7),
+            8  => read_drx!(self, dr8),
+            9  => read_drx!(self, dr9),
+            10 => read_drx!(self, dr10),
+            11 => read_drx!(self, bkp_dr11),
+            12 => read_drx!(self, bkp_dr12),
+            13 => read_drx!(self, bkp_dr13),
+            14 => read_drx!(self, bkp_dr14),
+            15 => read_drx!(self, bkp_dr15),
+            16 => read_drx!(self, bkp_dr16),
+            17 => read_drx!(self, bkp_dr17),
+            18 => read_drx!(self, bkp_dr18),
+            19 => read_drx!(self, bkp_dr19),
+            20 => read_drx!(self, bkp_dr20),
+            21 => read_drx!(self, bkp_dr21),
+            22 => read_drx!(self, bkp_dr22),
+            23 => read_drx!(self, bkp_dr23),
+            24 => read_drx!(self, bkp_dr24),
+            25 => read_drx!(self, bkp_dr25),
+            26 => read_drx!(self, bkp_dr26),
+            27 => read_drx!(self, bkp_dr27),
+            28 => read_drx!(self, bkp_dr28),
+            29 => read_drx!(self, bkp_dr29),
+            30 => read_drx!(self, bkp_dr30),
+            31 => read_drx!(self, bkp_dr31),
+            32 => read_drx!(self, bkp_dr32),
+            33 => read_drx!(self, bkp_dr33),
+            34 => read_drx!(self, bkp_dr34),
+            35 => read_drx!(self, bkp_dr35),
+            36 => read_drx!(self, bkp_dr36),
+            37 => read_drx!(self, bkp_dr37),
+            38 => read_drx!(self, bkp_dr38),
+            39 => read_drx!(self, bkp_dr39),
+            40 => read_drx!(self, bkp_dr40),
+            41 => read_drx!(self, bkp_dr41),
+            42 => read_drx!(self, bkp_dr42),
+            x  => panic!("Backup Data Register dr{} does not exist.", x),
```

Comment:
```suggestion
            x  => unreachable!(),
```

Reply:
I wonder why `unreachable!` is better than `panic!` with a descriptive error message?

Reply:
I believe you can provide unreachable with a message as well which makes it even more descriptive.

One problem with messages though, is that the string to be written has to be stored somewhere which might use valuable resources

Reply:
Oh, plenty reasons. Formatting is expensive, `unreachable!()` will just throw out a fixed string which can also be reused if used in multiple places. Also it tells the compiler (and the reader) that this situation can not happen under normal circumstances... You're really not expected to see that `panic!()` (and it's message) in embedded implementations, it's just to prevent undefined behaviour.

Reply:
Thanks for the clarification! I will change as suggested.